Issue #2546: added new check RegexpOnFilenameCheck

rnveach opened the pull request:
Start of Check.
No documentation until requirements are confirmed.

**Properties**:
match:  true/false. report matching or mismatching of folders and/or files by Patern, default is matching
folderPattern: regular expression. folder pattern to use, default is all folders regardless of match
fileNamePattern: regular expression. file name pattern to use, default is all files regardless of match

All properties are optional. If none are set, fileNamePattern defaults to "\s" and "match" defaults to true.

**Explanation**:
1) If users don't specify a file or folder pattern, it defaults to looking for spaces in file names of all folders.

2) If users don't specify a folder pattern, all files using the file name pattern is matched against.
If a file matches the pattern and match is true, it will report the file. If a file doesn't match the pattern and match is false, it will report the file.
Like the original example, this can be used for preventing file extensions anywhere: "{0} must be in PNG format, not GIF."
````
Example: Match all GIF files
match = true
fileNamePattern = ".*\\.gif"
````

3) If users don't specify a file pattern, all folders using the folder pattern is matched against.
If a folder matches the pattern and match is true, it will report all files in the folder. If a folder doesn't match the pattern and match is false, it will report all files in the folder.
This wasn't in the original example. This can be used to for prevent folder names, folders from appearing in certain locations, or files in a specific folder.
````
Example: Match all files that are in a directory with an uppercase letter
match = true
folderPattern = "[A-Z]"
````

4) If the users specify a folder and a file pattern:
A) All folders are looking for a match. (No mismatching allowed like without fileNamePattern)
B) If the folder matches, all files follow its normal matching/mismatching rules.

Like the original example, this can be used for requiring files in specific folders or specific names.
````
Example: Match all files in resource directory that don't start with "Input". (CS case)
match = false
folderPattern = "[\\/]src[\\/]\\w+[\\/]resources[\\/]"
fileNamePattern = "^Input"
````


For A):  I was originally going to go with its default way of matching/mismatching, but I am not sure if this made it too confusing to understand and I'm not sure I can come up with a valid use. Let me know if I should re-implement it.
````
Example: Match all non-properties files not in a "resource" directory. 
match = false
folderPattern = "[\\/]resource[\\/]"
fileNamePattern = "^.*\\.properties$"
````
With the way the code is implemented now, it matches all non-properties files in "resource" directories.

**Changes**:
> targetFolderPattern 

I renamed this to folderPattern to match fileNamePattern's name, since they are so similar.

**Other**:
Any ideas for a default message for "regexp.filepath.match" and "regexp.filepath.mismatch"?
Original forum has "File {0} does not match '{1}' pattern", which sorta works if we combine file pattern and folder pattern in "{1}".

mkordas commented:
Thanks for the prototype! Couple if initial comments from my side:
1. I would rename all occurrences of 'folder' to 'directory' as the latter seems to be [name of Wikipedia page on the topic](https://en.wikipedia.org/wiki/Directory_(computing)).
2. We cannot confuse ['matching' and 'finding'](http://stackoverflow.com/questions/4450045/difference-between-matches-and-find-in-java-regex) regex. In all names 'match' is used, while in the code I see call to 'find()'
3. Pattern `[\\/]` is crucial here as otherwise check isn't platform independent. I think we should remind users about it in docs
4. As we use `find()`, `.*` is obsolete in `fileNamePattern = ".*\\.gif"`
5. I don't quite understand `If a folder matches the pattern and match is true, it will report all files in the folder. If a folder doesn't match the pattern and match is false, it will report all files in the folder.`
6. 
> With the way the code is implemented now, it matches all non-properties files in "resource" directories.

It is not intutitive that `match = false` relates to `fileNamePattern`, while `folderPattern` remains positive match. We need to rethink this.

rnveach commented:
> 5. I don't quite understand

If it helps, my item 3 is the same as item 2, the only difference is one applies to the folders while the other applies to the fileNames. I was just trying to explain the difference between match being true and false, when a fileNamePattern isn't set and only the folderPattern is set.
If not maybe the examples will be simpler to understand?
````
Example: Match all files in directories where the directory has an uppercase letter
match = true
folderPattern = "[A-Z]"

Example: Match all files in directories where the directory has a space or special character (ie not '\\w').
match = false
folderPattern = "\\w"
````

mkordas commented:
@rnveach 
With examples it is clear. We will need a lot of examples in docs anyway. Some ideas:
* no files matching `Tests$` - most probably they were supposed to be `Test`
* in `src/main/java` just `.java$` files, in `src/main/groovy` just `.groovy$` files - very common mistake in my projects, surprisingly build works even when languages are mixed :)
* you already covered: no whitespaces, whitelist of allowed characters: `[0-9a-zA-Z-.,;_]` - [source](http://superuser.com/a/748264/474103)

rnveach commented:
@mkordas It is not that I don't appreciate extra examples, but lets focus on solidifying the requirements and properties of the check because things may change like....

I now see an issue with the way folderPattern will work.
> Example: Log all files in directories where the directory has an uppercase letter

Since it is running match on the entire directory structure, this example will cause issues with Windows users because all drive letters start with an uppercase letter.
Also because it is the entire directory structure, we would give errors on folders the project is stored in, in addition to folders inside the project. Ex: `/home/Ricky/checkstyleWorkspace/src/...`

> 6 . It is not intutitive that match = false relates to fileNamePattern, while folderPattern remains positive match. We need to rethink this.

And we haven't solved this issue, unless we just have them rely on the documentation to explain the behavior.

We have only 2 choices I believe:
1) Remove the option of matching folders alone and stick to just matching files, as we may be trying to put too much into this one check and could be in another check. This removes my original option 3 (from my first post), and makes the default behavior the same as option 1. This would solve your issue 6.
2) Add a new property "baseDirectory" and require the user to specify their base directory, since we can't generate/capture it automatically right now, so we can remove it before doing any folder matching. This would not solve your issue 6. Since its not automatic, this could cause issues on with different bases on different computers and operating systems.

mkordas commented:
> It is not that I don't appreciate extra examples, but lets focus on solidifying the requirements and properties of the check because things may change like

@rnveach I think the examples of usage are great way to discuss requirements. If we are able to provide working examples for multiple real-life use cases, then it means that the check is useful and there are no major gaps in requirements 

> Remove the option of matching folders alone and stick to just matching files, as we may be trying to put too much into this one check

Agreed, let's limit functionality for this check

@rnveach, will this check be suitable to validate file naming conventions? For example I'd like to have all XML files to be named in camelCase only.

rnveach commented:
> I'd like to have all XML files to be named in camelCase only.

Since we can't remove the file extension from the fileNamePattern, we have to have a full true/false match on the entire name. Since you want a true match on 'xml', the rest of the pattern has to be a true match too.
By camel case, you mean to flag all files that either: 1) start with an uppercase letter, 2) have 2 or more uppercase letters in a row, 3) start with an uppercase letter after a period, or 4) have a space (or any special character you wish).
Your `fileNamePattern` would be `((^[A-Z])|([A-Z][A-Z]+)|(\.[A-Z])|( )).*\.xml$` and `match` would be `true`. It should work on files with double periods too.
If you wanted lowercase letters to be more than 1 character before an uppercase character, this would make it harder.

We could add an property to drop file extensions from the fileNamePattern for complicated expressions. Since it is an `AbstractFileSetCheck` we have the property `fileExtensions` to supply true matches on file extensions, so I don't see this kind of property causing any issues.
This would allow us to do false matches on just the actual file name, so you could do an easier, and more common, expression like `^([a-z]+([A-Z][a-z0-9]+)*\.?)*$` with `match` be `false` and something like `ignoreFileExtension` be `true`. It should work on files with double periods too, hence the '\.?' part.

> Agreed, let's limit functionality for this check

I'll make updates soon so you can test the examples out if you wish.

tsjensen commented:
Just to let you know: The *RegexpOnFilename* check was originally my idea (as documented [here](https://groups.google.com/forum/#!topic/sevntu-checkstyle/W1Kt3K1-8R8)). I [published the check independently](http://checkstyle-addons.thomasjensen.com/latest/checks/regexp.html#RegexpOnFilename) more than half a year ago.
You may of course do your own reimplementation for Checkstyle, as you did. Just be sure to give appropriate credit for the idea. Thank you.

romani commented:
@rnveach , are you blocked by smth to proceed ? Let me know If I can help you.

rnveach commented:
@romani Sorry for the delay.

I think it is ready for review.
I filled in the documentation. Let me know if the xdocs need more.

Is it ok that I catch IOException from `File.getCanonicalPath`? This is how the original implementation was coded. Not sure if a checkstyle exception should be thrown instead.
If it is ok, I will add an override to idea's warning.
I will fix the other warnings in the next update.

tsjensen commented:
I find no reference to the fact that you copied the idea for this check from Checkstyle Addons. A remark along the lines of *"Created after the check of the same name from [Checkstyle Addons](http://checkstyle-addons.thomasjensen.com/latest/checks/regexp.html#RegexpOnFilename)"* in the docs would be fine for me. What are your plans for this?

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.java:
```diff
@@ -0,0 +1,179 @@
+            folderPath = file.getParentFile().getCanonicalPath();
+        }
+        catch (IOException ex) {
+            filePath = file.getAbsolutePath();
```
@rnveach we shouldn't ignore any exceptions unless absolutely necessary. What are cases when absolute path works but canonical path fails?

rnveach replied:
@mkordas 
`RegexpOnFilenameCheckTest.testException` showcases when this happens.
Truthfully, I don't think this exception can be reached normally since we have already read the file by this point. The main reason for the catch is because we can't throw the IOException as is, and like I said I just used the original implementation's logic for that.

mkordas commented on src/xdocs/config_regexp.xml:
```diff
@@ -554,6 +554,134 @@
+          To configure the check to report file names that contain a space:
+        </p>
+        <source>
+&lt;module name=&quot;RegexpOnFilename&quot;/&gt;
```
@rnveach We need to explain default behaviour also in first sentences

mkordas commented on src/xdocs/config_regexp.xml:
```diff
@@ -554,6 +554,134 @@
+            <td>fileNamePattern</td>
+            <td>Regular expression to match the file path against.</td>
+            <td><a href="property_types.html#regexp">regular expression</a></td>
+            <td><code>null</code></td>
```
@rnveach is it really `null` here?

rnveach replied:
@mkordas 
We don't set the pattern fields when the class is created. They are only filled in once the `init` method is called if the fields are still null, meaning the user didn't specify any properties. This is by design so when the users specify atleast one property, all the matching are under their control for full customization and not defaulting to the 'no-space' version.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheckTest.java:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    @Test
+    public void testIt() throws Exception {
```
@rnveach It could be named `testDefaultConfigurationOnValidInput`

mkordas commented on config/suppressions.xml:
```diff
@@ -74,7 +74,7 @@
 
     <!-- Tone down the checking for test code -->
-    <suppress checks="CyclomaticComplexity" files="[\\/]XDocsPagesTest\.java" lines="298"/>
+    <suppress checks="CyclomaticComplexity" files="[\\/]XDocsPagesTest\.java" lines="299"/>
```
@rnveach can you explain reason of that change?

rnveach replied:
@mkordas 
The new check was added to `XDocsPagesTest.XML_FILESET_LIST` for proper validation during the XDoc tests, like the other `Regexp` checks listed there. The suppression is there because one of the methods in the class was never fixed for how complicated it is.

romani commented:
@tsjensen ,

>I find no reference to the fact that you copied the idea for this check from Checkstyle Addons. A remark along the lines of "Created after the check of the same name from Checkstyle Addons" in the docs would be fine for me. What are your plans for this?

We already had that conversation with you. Nothing is changed from that time. We will not give any reference to you in code.  Commit will have reference to issue, in issue there are marks who proposed this Check (you) and all link to all our discussion about different views on design , and your public reject to contribute to our project in the form that we want.
 
We do not reference in code persons who give idea, we never did , we will never do. In future releases we even remove all `@author` tags, as they are just polluting the code and are not up-to-date and ..... . Recognition page is https://github.com/checkstyle/checkstyle/graphs/contributors

I could give you special recognition only if you have patent of this algorithm. This will make me sure that this is completely your idea, and not an idea of your friend, co-worked, mother, some man question on stackoverflow ..... .

tsjensen commented:
@romani As you know, I am not talking about `@author` tags, as I am not the author of your code. But still the idea is mine and I *am* the author of the implementation and documentation of the check in Checkstyle Addons, which has existed for a long time before you started this one. So I think stealing the idea without any visible mention of its origin would be plagiarism.

romani commented:
I already gave you full answer.

romani commented:
@rnveach , CIs has to be green, before I start code review.

from CIs:
> [checkstyle] [ERROR] /home/travis/build/checkstyle/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheckTest.java:226: Unicode escape(s) usage should be avoided. [AvoidEscapedUnicodeCharacters]

please try to avoid of Unicode, I am not completely sure a reason of test , but if want to have wrong filepath, there are other ways to do it. If you can explain why you need such unicode , please add this case to suppressions.

> regexp.filepath.match=TODO ''{0}''

Please do a message.

mkordas commented on src/xdocs/config_regexp.xml:
```diff
@@ -554,6 +554,134 @@
+&lt;/module&gt;
+        </source>
+        <p>
+          To configure the check to only allow java or xml files in your folders:<br />
```
@rnveach Instead of `To configure the check to only allow java or xml files in your folders:` let's use `To configure the check to only allow Java or XML files in your folders use the below.`, because there's also note before the example.

rnveach commented:
@romani 
It is needed to test `File.getCanonicalPath` throwing an IOException on all OSs. I added a suppression.
http://stackoverflow.com/questions/4489878/file-getcanonicalpath-failure-examples#26412778

@mkordas 
I fixed your issues. The default behavior is described in the Check's description and not in the example.

romani commented:
@rnveach ,

1) please propose a non default configuration(s) to be useful in our project, we need to see by examples how useful this Check and what it can do, how easily its confguration to setup/read, ....

2) please start to write javadoc in plain text format (gist) to make sure we see all nuances of the Check .

rnveach commented:
@romani 
1) Done, added 3. I was going to run something over the resource folders to enforce "Input*.java" but since CS excludes them this isn't possible. We would have to suppress everything but this or create a whole new CS run for resources.
2) Done and printed site. http://rnveach.github.io/checkstyle/apidocs/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.html and http://rnveach.github.io/checkstyle/config_regexp.html#RegexpOnFilename

I added `allowIfAllCharactersEscaped` for `AvoidEscapedUnicodeCharacters` for the unicode in test and removed the suppression. I originally wanted to use `allowByTailComment` but we have another check that forbids tail comments and `AvoidEscapedUnicodeCharacters` only works with tail comments.

romani commented:
> I was going to run something over the resource folders to enforce "Input*.java" but since CS excludes them this isn't possible. We would have to suppress everything except this check, OR create a whole new CS run for resources.

lets skip this and be simple.

mkordas commented:
@rnveach 
> @mkordas 
I fixed your issues. The default behavior is described in the Check's description and not in the example.

Thanks for that!

romani commented:
> http://rnveach.github.io/checkstyle/apidocs/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.html

1) 
>  their proper folders and don't show up in the wrong ones

lets avoid "wrong" "proper",  we have in scope of this check only term - "match" "mismatch".
Additionally please add explanation of all options to the javadoc.

2)
> To configure the check to perform the same test above but only on files in the source folder:

please avoid reference to other examples, copy paste of explanation part from previous "allow Java and XML files" will not make this line longer

3)
>  To configure the check to only allow file names to be camel case: 

This example is too artificial and useless, we need more real example. Look at https://groups.google.com/d/msg/sevntu-checkstyle/W1Kt3K1-8R8/1DGROqOa9M4J where we discussed examples of Check usage with author of Check idea ( @tsjensen ). If we can not find any reasonable example for of usage for each option it is good sign that option is useless.
We need to have examples for each option.

http://rnveach.github.io/checkstyle/config_regexp.html#RegexpOnFilename

4)
I do not see in description an algorithm and sequence of matching by Regexp user can put in options. That need to be clearly described no secret logic, all should be transparent.

5)
> fileNamePattern 	Regular expression to match the file path against.

so file name or file path ?

6)
> fileExtensions 	file type extension of files to process

we need to describe here how it is related to other Patterns.

7)
The best way to test design of the Check is to give configuration examples to your work colleges and ask them to explain what this config will enforce. If user explain you all without reading documentation - that is really good design. Glance at html documentation helps the user it is ok.

https://github.com/rnveach/checkstyle/commit/306f04c9a3cc12ebbbdb26c58317c891ede7f951

8)
> src/xdocs/checks.xml

there should first line from javadoc, as it is done there http://rnveach.github.io/checkstyle/apidocs/com/puppycrawl/tools/checkstyle/checks/regexp/package-summary.html . First line of javadoc is special line.

9)
> File was found to have matched the illegal folder pattern ''{0}'' and file pattern ''{1}''.

I do not like "illegal" it is too subjective. We should be objective in message and state only why we show this message. Smth like "File match folder pattern ''{0}'' and file pattern ''{1}" is what Check should do. If user want more info he need to override message to be more descriptive why matching by this pattern is a problem  and how to resolve it.
Please do translation to all other languages by Google translator.


https://github.com/rnveach/checkstyle/blob/306f04c9a3cc12ebbbdb26c58317c891ede7f951/src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.java

10)
> `* @author CheckStyle Team`

please remove this, that is useless

11)
> Compiled regexp to match a file.
>  Setter for file name format.
> Sets whether the check should look for a file name match or mismatch.
file ? file name or full file path ? please be exact as possible.

12)
> String filePath;
        final String folderPath;

I know you like "final" .... in this case please do code smth like : 
final String filePath = getFileName(file, ignoreExtension);
final String folderPath = getFullPath(file);

13)
>         final String folder;
        if (folderPattern == null) {
            folder = "";
        }
        else {
            folder = folderPattern.toString();
        }

should be smth like:

final String folder = toStringOrDefault(folderPatern, "");

14)
>  private boolean isMatchFolder(String folderPath) {
> private boolean isMatchFile(String filePath) {

one return point please.

15)
>         if (fileNamePattern == null) {
            useMatch = match;
        }
        else {
            useMatch = true;
        }

please put explanation of this code to comment and jaavdoc, I do not think it is required.

16)
> private void log() {

move this method to the end, methods should appear in code by their appearance in the code from top to bottom. Sorry that this is not automated yet, but work is in progress for such Check in sevntu .

rnveach commented:
1) Done.
2) Done, added some more text to make it clearer on its purpose.
3)
I don't see how you can say this is useless. It will validate resource file names that other checks currently can't.
mkordas requested something similar to this in a previous post in this PR. Since it was already made, I copied it in as an example and just made it apply to more files.

I changed the expression to first uppercase as requested in chat.
Do you still not like this type of configuration?
5) Done and fixed in other areas I saw.
6) Done. The text that was there was copied from another check that implements the same property.
8) Done and made line shorter.
9) Done.
10) Changed to my name since JavadocType requires the tag.
11) File name as I corrected in 5.
12) Done.
13) Done.
14) Done.
16) Done.

romani commented:
1)
>```              fileNamePattern = CommonUtils.createPattern("\\s");
            match = true;
```

looks like second statement is useless, it is already done in fields initialization.


2)
>```
            final int extension = fileName.lastIndexOf('.');

            if (extension > 0) {
                fileName = fileName.substring(0, extension);
            }
```

http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/Files.html#getNameWithoutExtension%28java.lang.String%29

3)
> private static String getStringOrDefault(Pattern pattern, String defaultString) {

should appear in code after first usage. Please do it single exit-point method.

romani commented:
4)
> // null pattern means 'match' applies to the folder pattern matching

I do not see description of this logic in javadoc/xdoc , please quote it if it exists or please append it.

that is hard for my brain, I do not understand why we need that logic and not simply follow "match" option for both patterns. Lets do this by example, please provide me real examples where such logic is useful.

rnveach commented:
Second list:
1) Done.
2) Done.
3) Done.

>  I do not understand why we need that logic and not simply follow "match" option for both patterns.

Since there is only one match variable, both patterns would be forced on/off down that path. This means your looking for a file pattern in a specific folder OR a non file match in a non folder match. The 2nd option I don't see it as really being usuable.

````
folder      file       description
1 match     match      Report files that don't belong in your folders. (like 'Match all GIF files')
2 match     non-match  Report files that don't match your file naming standard in your folders (like 'Match all files in resource directory that don't start with "Input")
3 non-match match      Report files that are found in unknown/invalid folders. (See below)
4 non-match non-match  Report unknown/invalid files that are found in unknown/invalid folders.
````

2 and 3 wouldn't be possible with both folder and file being controlled by the one match variable.
1 and 2 have written examples and work in current implementation.
3 is works in the current implementation if we are reporting all files (no file pattern). It could be used report files in folders with uppercase letters.
````
Example: Log all files in directories where a directory in 'src' has a space or special character (ie not '\\w') or all files in directories that aren't 'src'.
match = false
folderPattern = "[\\/]src([\\/]\\w+)*$"
````
If 3 did allow file names for the pattern, we can use other examples like 'Report all property files that aren't in a resource folder.', but this would require 2 different match variables for file and folder or change match to allow 3 states.
4 isn't possible in the current implementation. I can't think of a good, real example of using this type of search right now.

romani commented:
third code review:

1)
>```  It is recommended that if you create your own pattern, to also specify a custom error message. This allows the error message printed to be clear what the violation is, especially if multiple RegexpOnFilename checks are used.

Argument 0 for the message populates the check's folderPattern. Argument 1 for the message populates the check's fileNamePattern. The file name is not passed as an argument since it is part of CheckStyle's default error messages. 
```

from http://rnveach.github.io/checkstyle/apidocs/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.html.
Please make join this paragraphs in one, as they explain the same topic. The same for Xdoc.

2)
> 4 isn't possible in the current implementation. I can't think of a good, real example of using this type of search right now

yes, it is weird. Now I see it. We need to explain this logic to users in documentation in clear way. They need to know how it works without any looking at code.

3)
> Log all files in directories where a directory in 'src' has a space or special character (ie not '\\w') or all files in directories that aren't 'src'.

I do not see a reason in such case, there will be always folders in project that is other then "src". 

4)
>  Report all property files that aren't in a resource folder.

good config, please add it to our examples set instead of example " Log all files in directories where a directory in 'src' has a space or special c ....".

rnveach commented:
Third list:
1) Done.
2) I added more description to the match property, and added a new paragraph in the description section. This should also mark 4 from the second list, and 15 from the first list as done.
3) The regular expression could be changed to make it accept other folders like 'src', I was just saying that is how it would work as it is and would need to be customized more for a specific use.
4) Done.

romani commented:
http://rnveach.github.io/checkstyle/config_regexp.html#RegexpOnFilename
http://rnveach.github.io/checkstyle/apidocs/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.html

1)
> ignoreFileExtensions
fileExtensions

this options looks like related as they are named the same, but this is not true.
Please rename `ignoreFileExtensions` to `ignoreFileNameExtensions` to be closer to `fileNamePattern`.

2)
> If this check is configured with no properties

Please move it as third paragraph. User need description first , and only after that all nuances with messages and defaults.

3)
> The match property controls whether the check reports files that match or don't match the supplied regular expressions. 

In this paragraph please describe the whole logic(sequence of matches) of how RegExp are applied and `fileExtensions` and `ignoreFileNameExtensions`

4)
> To configure the check to only allow property files to be located in the resource folder

please make it more close to real life and allow xml files too.

rnveach commented:
Fourth list:
1) Done.
2) Done.
3) Done.
4) Done.

romani commented:
https://github.com/rnveach/checkstyle/blob/b8119f9285ff63a9a388d5ab6a0afd5e4fd085d5/src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheck.java

all changes need to be done in xdoc too.

1)
> Implementation of a check that file names .....

I do not understand this sentence . Please rephrase it.

2)
> The file extensions property first ignores files that don't match any of the specific extensions supplied.

"file extensions" ??? there is "fileExtension" property. All other references to property should have exact name.
"ignores ... don't match" - please rephrase to use positive words.

3)
>  To configure the check to only allow property and xml files to be located in

this need to be second example, before examples with Patterns.

4)
> Note: 'folderPattern' must be specified if checkstyle is analyzing

move it to next configuration example. No need to explain nuances in advance.

5)
> The match property controls whether the check reports files that match or don't match the supplied regular expressions.

Please explain whole algiorithm by words. This sentence is strange and out of context.

6)
Please give description to your any co-worker and ask him how much is it clear  for him.

rnveach commented:
Fifth list:
1 & 5) Better?
2) Done
3) Done
4) Done

romani commented:
merged as FF.

thanks a lot for your hard work.
